hoomd/mpcd/ParticleData.cc

```diff
@@ -416,6 +457,128 @@ void mpcd::ParticleData::initializeRandom(unsigned int N, const BoxDim& local_bo
         }
     }
 
+/*!
+ * \param density Density of particles inside sphere
+ * \param R Radius of sphere 
+ * \param local_box Local simulation box
+ * \param kT Temperature (in energy units)
+ * \param seed Random seed
+ * \param ndimensions Dimensionality
+ */
+void mpcd::ParticleData::initializeRandomSphere(Scalar density, Scalar R, const BoxDim& local_box, Scalar kT, unsigned int seed, unsigned int ndimensions)
+    {
+    // only one particle type is supported for this construction method
+    m_type_mapping.clear();
+    m_type_mapping.push_back("A");
+    // default particle mass is 1.0
+    m_mass = Scalar(1.0);
+
+    // throw runtime error if radius is negative or zero
+    if (R <= Scalar(0.0))
+        {
+        throw std::runtime_error("Radius of sphere can't be negative or zero");
+        }
+
+    // Calculating Ndraw - number of particles that should be inside the local box according to density
+    unsigned int Ndraw = round(density*local_box.getVolume());
+
+    // Allocating an Array for storing the random drawn positions
+    GPUArray<Scalar4> pos_draw(Ndraw, m_exec_conf);
+    m_pos_draw.swap(pos_draw);
+    ArrayHandle<Scalar4> h_pos_draw(m_pos_draw, access_location::host, access_mode::overwrite);
+
+    // center of box
+    const Scalar3 lo = local_box.getLo();
+    const Scalar3 hi = local_box.getHi();
+
+    // random number generator
+    std::mt19937 mt(seed);
+    std::uniform_real_distribution<Scalar> pos_x(lo.x, hi.x);
+    std::uniform_real_distribution<Scalar> pos_y(lo.y, hi.y);
+    std::uniform_real_distribution<Scalar> pos_z(lo.z, hi.z);
+    std::normal_distribution<Scalar> vel(0.0, fast::sqrt(kT / m_mass));
+
+    // generating Ndraw random positions and keeping them if they lie inside sphere(also counting N_kept)
+    unsigned int N_kept = 0;
+    unsigned int N_tag_rank = 0;
+
+    for (unsigned int i=0; i < Ndraw; ++i)
+        {
+        Scalar pos_x_draw = pos_x(mt);
+        Scalar pos_y_draw = pos_y(mt);
+        Scalar pos_z_draw = (ndimensions == 3) ? pos_z(mt) : Scalar(0.0);
+        Scalar mod_pos_draw = slow::sqrt(pos_x_draw*pos_x_draw + pos_y_draw*pos_y_draw + pos_z_draw*pos_z_draw);
+        if (mod_pos_draw < R)
+            {
+            h_pos_draw.data[N_kept] =  make_scalar4(pos_x_draw,
+                                                    pos_y_draw,
+                                                    pos_z_draw,
+                                                    __int_as_scalar(0));
+            N_kept += 1;
+            }
```

Comment:
This could be written more cleanly like this:
```suggestion
        const Scalar3 pos = make_scalar3(pos_x(mt), pos_y(mt), (ndimensions == 3) ? pos_z(mt) : Scalar(0.0));
        const Scalar r = slow::sqrt(dot(pos, pos));
        if (r < R)
            {
            pos_draw.push_back(make_scalar4(pos.x, pos.y, pos.z, __int_as_scalar(0));
            }
```